### PR TITLE
Add possibility to run Python scripts with current interpreter

### PR DESCRIPTION
**Pull request by jpakkane**
When using the MSI installer version, Python might not be available on the current system (or it might be the wrong version). Thus add functionality to run Python scripts with the embedded interpreter.

This is not finished yet. I need to fix all places where `sys.executable` is being used. Just putting this here to run through CI first and to gather comments.

### CODE REVIEW

**Comment by polarina** [comment]
What kind of scripts do you expect someone to execute through `meson runpython`?

**Comment by jpakkane** [comment]
It is both for internal use (many internal scripts are written in Python) and for scripts people want to run with the Python 3 module's `find_python` command. Also `find_program` will be updated to detect Python scripts and run them with the internal interpreter so builds needing Python will work even on Windows when running from the .msi packages and Python is not installed.

**Comment by dcbaker** [comment]
One comment, how will this work with 3rd party modules? Since they can't be added to the internal python implementation, I think it would be best to have this interface only be available to internal scripts, and require python3 to be installed separately for users wanting to use python in their project.

**Comment by jpakkane** [comment]
Or we could check if the system has an installed Python 3 in path and use that if it exists only using the internal interpreter if it is not available.

**Comment by rhd** [comment]
I'm a huge +1 at this PR. I'm maintaining a windows and Linux build and I'm writing bash (for Linux) and batch (for windows). It sucks. This would allow me to write one script that works everywhere.

File: mesonbuild/backend/vs2010backend.py [hunk 1]
```diff
@@ -396,10 +396,11 @@ def gen_run_target_vcxproj(self, target, ofname, guid):
         action = ET.SubElement(root, 'ItemDefinitionGroup')
         customstep = ET.SubElement(action, 'PostBuildEvent')
         cmd_raw = [target.command] + target.args
-        cmd = [sys.executable, os.path.join(self.environment.get_script_dir(), 'commandrunner.py'),
-               self.environment.get_build_dir(),
-               self.environment.get_source_dir(),
-               self.get_target_dir(target)] + self.environment.get_build_command()
+        cmd = mesonlib.python_command +
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[Flake8]__

SyntaxError: invalid syntax

File: mesonbuild/backend/vs2010backend.py [hunk 2]
```diff
@@ -396,10 +396,11 @@ def gen_run_target_vcxproj(self, target, ofname, guid):
...
+            [os.path.join(self.environment.get_script_dir(), 'commandrunner.py'),
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[Flake8]__

unexpected indentation

File: mesonbuild/environment.py [hunk 3]
```diff
@@ -365,14 +365,10 @@ def get_coredata(self):
...
-        # Python might not be installed so don't prefix
-        # the command with it.
-        if sys.executable.endswith('meson.exe'):
-            return [sys.executable]
-        if unbuffered:
-            [sys.executable, '-u', self.meson_script_launcher]
-        return [sys.executable, self.meson_script_launcher]
+        cmd = mesonlib.meson_command[:]
+        if unbuffered and 'python' in cmd[0]:
+            cmd = [cmd[0],'-u'] + cmd[1:]
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[Flake8]__

missing whitespace after ','

File: mesonbuild/scripts/regen_checker.py [hunk 4]
```diff
@@ -32,15 +32,11 @@ def need_regen(regeninfo, regen_timestamp):
...
-    if sys.executable.lower().endswith('meson.exe'):
-        cmd_exe = [sys.executable]
-    else:
-        cmd_exe = [sys.executable, mesonscript]
-    cmd = cmd_exe + ['--internal',
-           'regenerate',
-           regeninfo.build_dir,
-           regeninfo.source_dir,
-           '--backend=' + backend]
+    cmd = mesonlib.meson_command + ['--internal',
```
Comment (ignatenkobrain) [inline mesonbuild/scripts/regen_checker.py]: __[Flake8]__

undefined name 'mesonlib'

File: msi/createmsi.py [hunk 5]
```diff
@@ -17,6 +17,7 @@
 import sys, os, subprocess, shutil, uuid
 from glob import glob
 import platform
+import shutil
```
Comment (ignatenkobrain) [inline msi/createmsi.py]: __[Flake8]__

redefinition of unused 'shutil' from line 17

File: mesonbuild/environment.py (hunk 3 above)
Comment (keszybz) [inline mesonbuild/environment.py]: `cmd.insert(1, '-u')`?

File: mesonbuild/mesonlib.py [hunk 6]
```diff
@@ -22,6 +22,15 @@
 
 from glob import glob
 
+if sys.executable.endswith('meson.exe'):
```
Comment (keszybz) [inline mesonbuild/mesonlib.py]: Here and in other places, `endswith` seems fragile. What about `os.path.basename(sys.executable) == 'meson.exe'` instead?

File: mesonbuild/environment.py [hunk 7]
```diff
@@ -365,14 +365,10 @@ def get_coredata(self):
...
-        # If running an executable created with cx_freeze,
-        # Python might not be installed so don't prefix
-        # the command with it.
-        if sys.executable.endswith('meson.exe'):
-            return [sys.executable]
-        if unbuffered:
-            [sys.executable, '-u', self.meson_script_launcher]
-        return [sys.executable, self.meson_script_launcher]
+        cmd = mesonlib.meson_command[:]
+        if unbuffered and 'python' in cmd[0]:
```
Comment (keszybz) [inline mesonbuild/environment.py]: What about `os.path.basename(cmd[0]).startswith('python')` instead?

File: mesonbuild/mesonlib.py [hunk 8]
```diff
@@ -22,6 +22,15 @@
...
+    # In Windows and using the MSI installed executable.
+    meson_command = [sys.executable]
+    python_command = [sys.executable, 'runpython']
+else:
+    meson_command = [sys.executable, os.path.join(os.path.split(__file__)[0], '..', 'meson.py')]
```
Comment (keszybz) [inline mesonbuild/mesonlib.py]: `os.path.split(__file__)[0]` is `os.path.dirname(__file__)`.

File: run_tests.py [hunk 9]
```diff
@@ -127,18 +133,28 @@ def get_fake_options(prefix):
...
         sys.stdout = old_stdout
         sys.stderr = old_stderr
     return returncode, mystdout.getvalue(), mystderr.getvalue()
 
+def run_configure_external(full_command):
+    pc, o, e = mesonlib.Popen_safe(full_command)
+    return pc.returncode, o, e
+
+def run_configure(meson_command, commandlist):
+    global meson_exe
```
Comment (keszybz) [inline run_tests.py]: `global` definition is not needed, variable is only read.

File: mesonbuild/backend/vs2010backend.py [hunk 10]
```diff
@@ -396,10 +396,11 @@ def gen_run_target_vcxproj(self, target, ofname, guid):
         action = ET.SubElement(root, 'ItemDefinitionGroup')
         customstep = ET.SubElement(action, 'PostBuildEvent')
         cmd_raw = [target.command] + target.args
-        cmd = [sys.executable, os.path.join(self.environment.get_script_dir(), 'commandrunner.py'),
-               self.environment.get_build_dir(),
-               self.environment.get_source_dir(),
-               self.get_target_dir(target)] + self.environment.get_build_command()
+        cmd = mesonlib.python_command + \
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[Flake8]__

undefined name 'mesonlib'

File: mesonbuild/mesonmain.py [hunk 11]
```diff
@@ -303,6 +303,12 @@ def run(args, mainfile=None):
             return mconf.run(remaining_args)
         elif cmd_name == 'wrap':
             return wraptool.run(remaining_args)
+        elif cmd_name == 'runpython':
+            import runpy
+            script_file = remaining_args[0]
+            sys.argv[1:] = remaining_args[1:]
+            runpy.run_path(script_file, run_name='__main__')
+            sys.exit(0)
```
Comment (rhd) [inline mesonbuild/mesonmain.py]: Does this mean that scripts that get run with the internal interpreter will always return 0 (ie: success)?

Comment (jpakkane) [reply mesonbuild/mesonmain.py]: Yes, unless they throw an exception or sys.exit(nonzero_number). Which is what they need to do with standalone scripts anyway.

Comment (rhd) [reply mesonbuild/mesonmain.py]: @jpakkane perfect. :+1:

**Pull request merged**